RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc

```diff
@@ -0,0 +1,742 @@
+void GlobalParticleTransformerAK8TagInfoProducer::fillParticleFeatures(DeepBoostedJetFeatures &fts,
+                                                                       const reco::Jet &jet) {
+  // require the input to be a pat::Jet
+  const auto *patJet = dynamic_cast<const pat::Jet *>(&jet);
+  if (!patJet) {
+    throw edm::Exception(edm::errors::InvalidReference) << "Input is not a pat::Jet.";
+  }
+
+  // do nothing if jet does not have constituents
+  if (jet.numberOfDaughters() == 0)
+    return;
+
+  // some jet properties
+  math::XYZVector jet_dir = jet.momentum().Unit();
+  GlobalVector jet_ref_track_dir(jet.px(), jet.py(), jet.pz());
+  const float etasign = jet.eta() > 0 ? 1 : -1;
+
+  TrackInfoBuilder trkinfo(track_builder_);
+
+  std::map<reco::CandidatePtr::key_type, float> puppi_wgt_cache;
+  auto puppiWgt = [&](const reco::CandidatePtr &cand) {
+    const auto *pack_cand = dynamic_cast<const pat::PackedCandidate *>(&(*cand));
+    const auto *reco_cand = dynamic_cast<const reco::PFCandidate *>(&(*cand));
+    float wgt = 1.;
+    if (pack_cand) {
+      wgt = pack_cand->puppiWeight();
+    } else if (reco_cand) {
+      if (use_puppi_value_map_) {
+        wgt = (*puppi_value_map_)[cand];
+      } else {
+        throw edm::Exception(edm::errors::InvalidReference) << "Puppi value map is missing";
+      }
+    } else {
+      throw edm::Exception(edm::errors::InvalidReference) << "Cannot convert to either pat::PackedCandidate or "
+                                                             "reco::PFCandidate";
+    }
+    puppi_wgt_cache[cand.key()] = wgt;
+    return wgt;
+  };
+
+  std::vector<reco::CandidatePtr> cpfPtrs, npfPtrs;
+  std::map<reco::CandidatePtr::key_type, bool> isLostTrackMap;
+
+  for (const auto &dau : jet.daughterPtrVector()) {
+    // remove particles w/ extremely low puppi weights
+    // [Note] use jet daughters here to get the puppiWgt correctly
+    if ((puppiWgt(dau)) < min_puppi_wgt_)
+      continue;
+    // from here: get the original reco/packed candidate not scaled by the puppi weight
+    auto cand = pfcands_->ptrAt(dau.key());
+    // charged candidate selection (for Higgs Interaction Net)
+    if (!include_neutrals_ && (cand->charge() == 0 || cand->pt() < min_pt_for_track_properties_))
+      continue;
+    // only when computing the nagative tagger: remove charged candidates with high sip3d
+    if (flip_ip_sign_ && cand->charge()) {
+      trkinfo.buildTrackInfo(&(*cand), jet_dir, jet_ref_track_dir, *pv_);
+      if (trkinfo.getTrackSip3dSig() > max_sip3dsig_)
+        continue;
+    }
+    if (cand->charge() != 0) {
+      cpfPtrs.push_back(cand);
+      isLostTrackMap[cand.key()] = false;
+    } else {
+      npfPtrs.push_back(cand);
+    }
+  }
+  // lost tracks: fill to cpfcands
+  for (size_t i = 0; i < lts_->size(); ++i) {
+    auto cand = lts_->ptrAt(i);
+    if (reco::deltaR(*cand, jet) < jet_radius_) {
+      cpfPtrs.push_back(cand);
+      isLostTrackMap[cand.key()] = true;
```

Comment:
@colizz Is `cand.key()` unique for lost tracks and packed candidates? Or the `key()` is just the index in each collection (then different types of objects can have the same keys)? Could you please check?

Reply:
hi @hqucms , they are unique keys for `reco::CandidatePtr` hence should be different for each lost track and packed candidates. 
<details>
  <summary>Here is my check:</summary>
```
cpf key: 37
cpf key: 39
cpf key: 42
cpf key: 46
cpf key: 249
cpf key: 1004
cpf key: 1012
cpf key: 1013
cpf key: 1014
cpf key: 1016
cpf key: 1017
cpf key: 1018
cpf key: 1019
cpf key: 1020
cpf key: 1024
cpf key: 1025
cpf key: 1028
cpf key: 1029
cpf key: 1030
cpf key: 1032
cpf key: 1036
cpf key: 1039
cpf key: 1040
cpf key: 1041
cpf key: 1042
cpf key: 1044
cpf key: 1051
cpf key: 1052
cpf key: 1304
cpf key: 44
cpf key: 50
cpf key: 51
cpf key: 1022
cpf key: 1023
cpf key: 1026
cpf key: 1027
cpf key: 1031
cpf key: 1033
cpf key: 1034
cpf key: 1035
cpf key: 1037
cpf key: 1038
cpf key: 1043
cpf key: 1045
cpf key: 1046
cpf key: 1049
cpf key: 1056
cpf key: 1069
cpf key: 1219
lts key: 16
lts key: 17
lts key: 47
=== end of jets ===
cpf key: 47
cpf key: 56
cpf key: 978
cpf key: 1001
cpf key: 1003
cpf key: 1005
cpf key: 1053
cpf key: 1055
cpf key: 1057
cpf key: 1058
cpf key: 1059
cpf key: 1060
cpf key: 1061
cpf key: 1062
cpf key: 1063
cpf key: 1064
cpf key: 1065
cpf key: 1066
cpf key: 1068
cpf key: 1134
cpf key: 57
cpf key: 998
cpf key: 1067
cpf key: 1071
cpf key: 1072
cpf key: 1073
cpf key: 1074
cpf key: 1075
cpf key: 1076
cpf key: 1048
cpf key: 1054
cpf key: 1070
cpf key: 387
lts key: 7
lts key: 10
lts key: 12
lts key: 13
lts key: 14
lts key: 19
lts key: 49
=== end of jets ===
```
</details>

Reply:
thanks a lot @hqucms , this is very nice catch. There is an overlap between "keys" in PF candidates and lost tracks, with a prob of ~0.5% (though small, it indicates a logical error in the code).
We should point out that in training GloParT, we used the same code to generate training samples, which means that a small portion of charged PF candidates were mistakenly labelled as lost tracks. For now, it appears that this error is consistent across both, allowing our trained GloParT model to perform the same in CMSSW inference.

It seems our best option is to retain this code to preserve GloParT's performance. Perhaps we could add a comment here for clarity. What do you think?

Reply:
@colizz Given the small possibility of the overlap, I would expect fixing this should have a rather negligible effect on the performance? If that is the indeed case I think it's better to fix it now already.

Reply:
ok, we will check whether the effect is negligible if we change the code here.

Reply:
hi @hqucms , the code has been fixed to use `CandidatePtr` itself as the map key: https://github.com/cms-sw/cmssw/pull/46523/commits/b813fb6bcb21869258145de96d49fa20136bc613
Additionally, `puppi_wgt_cache` had a similar issue; however, testing revealed that it does not produce the same errors as `isLostTrackMap`. In this updated code, we validated the scores and found that in <1% of cases, some charged PF cands are marked as `isLostTrack=True`, but this has only a minimal impact on the GloParT output scores. We also examined the ROC curve on a large sample set, and found no change to the ROC performance. Thus we conclude that this modification has a negligible impact on GloParT performance

Reply:
Thanks a lot @colizz for the fast turnaround! The fix looks good to me.